### DISCUSSION TITLE
401 Unauthorized error in Species Distribution

### DISCUSSION THREAD

**Discussion by eculler** [earthlab-education/Earth-Analytics-AY24#517] [Q&A]
Hi folks,

My part 3 video is currently delayed because I ran into an issue, which I was able to resolve. I'm posting this in case anyone else runs into it!

TL;DR -- Sign up for your GBIF account using a new username and password, NOT by using your GitHub account!

I found that when I signed up with GitHub, a couple things happened:

1. I had to reset the password -- I'm not sure what it was using for a password, but you need a password to use `pygif` and the API.
2. Even then, I got the following error: "Unauthorized with error status code 401 check your number of active downloads.", despite the fact that there were no active downloads on the account.

I was able to resolve this issue by creating a new account using a username and password instead of GitHub. You can link your account to GitHub later on if you like -- I've had no trouble with that!

Stay tuned for the corrected video later today! And if you don't finish the portfolio post by Tuesday evening, don't worry about it for now. We have a week in November scheduled for getting caught up with portfolio posts. Hopefully NCEI will be back up post hurricane too.

**Comment by hanried** [comment]
Hi @eculler! Thanks for posting about this - I'm also running into that error when I try to run the code to download the data from GBIF, however I am using a different email and password than my GitHub account. I tried stopping and restarting my codespace and that didn't help. Any thoughts on why the error is still happening? Thanks in advance!

**Comment by hanried** [comment]
@ZimmZone and @ZacharyCJDecker, this is how Elsa and I got my code to work in the species-distribution-98-download notebook for the "Download data from GBIF" section. I think this is the section we were talking about at the beginning of class today. I was also getting the 401 error but it wasn't because of my email, I think it was something to do with the actual download. Else helped me get this working during office hours, hopefully it helps!

1. After I ran the original code and got the 401 error, we went to the downloads section of my account on gbif.org and clicked on where it says DOI:
<img width="593" alt="image" src="https://github.com/user-attachments/assets/bbf73bd4-99a1-44c1-be10-f147d2f3c511">

2. On the DOI site, the last part of the URL is the string for the os.environ['GBIF_DOWNLOAD_KEY'] code, the very first line. To be totally honest, I don't remember what all else we had to change in the code after that first line, but hopefully you can compare with your code and get your download to work.
<img width="413" alt="image" src="https://github.com/user-attachments/assets/995e4f7b-3da2-47ad-ac27-f89e7a4cab45">


```
os.environ['GBIF_DOWNLOAD_KEY'] = '0003394-241007104925546

# Only download blue jay data once
gbif_pattern = os.path.join(gbif_dir, '*.csv')
if not glob(gbif_pattern):
    # Submit query to GBIF
    # Only download once
    if not 'GBIF_DOWNLOAD_KEY' in os.environ:
        gbif_query = occ.download([
            "speciesKey = 2482593",
            "year = 2023",
            "hasCoordinate = TRUE",
        ])
        download_key = gbif_query[0]
        os.environ['GBIF_DOWNLOAD_KEY'] = gbif_query[0]
    else: 
        download_key = os.environ['GBIF_DOWNLOAD_KEY']

    # Wait for the download to build
    wait = occ.download_meta(download_key)['status']
    while not wait=='SUCCEEDED':
        wait = occ.download_meta(download_key)['status']
        time.sleep(5)

    # Download GBIF data
    download_info = occ.download_get(
        os.environ['GBIF_DOWNLOAD_KEY'], 
        path=data_dir)

    # Unzip GBIF data
    with zipfile.ZipFile(download_info['path']) as download_zip:
        download_zip.extractall(path=gbif_dir)

# Find the extracted .csv file path
gbif_path = glob(gbif_pattern)[0]
```

**Reply by ZimmZone** [reply]
Thanks @hanried . It turns out mine was saying that I had too many downloads (I didn't have any) because what it meant was that I was not authorized to download. It looks like I had entered my credentials incorrectly so it wouldn't let me download. Once I fixed that I was good to go! 

Thanks for posting all of this - I really appreciate the help.

**Reply by hanried** [reply]
Ah nice, glad you got it working!!

**Comment by byandell** [comment]
I am still stuck. I created a new account (yandell.brian) and got the same issue. No progress. B
